### PR TITLE
Improve third-party Astro package support

### LINKED ISSUE #4071: 🐛 BUG: astro-embed integration fails due to .astro file extension error
### What version of `astro` are you using?

1.0.0-rc.1

### Are you using an SSR adapter? If so, which one?

None

### What package manager are you using?

npm

### What operating system are you using?

Mac (macOS Monterey 12.4)

### Describe the Bug

Site start in dev mode fails with astro-embed integration enabled in astro.config.mjs with warning about unknown .astro file extension type.

```
[plugin:astro] Unknown file extension ".astro" for /Volumes/BFD/Sites/astrobug/node_modules/@astro-community/astro-embed-twitter/Tweet.astro

You likely need to add this package to `vite.ssr.noExternal` in your astro config file.
```

### Link to Minimal Reproducible Example

https://github.com/scottaw66/astrobug

### Participation

- [ ] I am willing to submit a pull request for this issue.
Screen Shot 2022-07-27 at 8 10 50 AM

### PR DESCRIPTION
**Pull request by delucis**
## Changes

Fix #4071 

Search for Astro package dependencies beyond a project’s direct dependencies.

See [this comment](https://github.com/withastro/astro/issues/4071#issuecomment-1236431134) for more background — it lays out some alternative approaches, I’d be curious what people think!

**TL;DR** Current logic only adds _direct_ dependencies to `vite.ssr.noExternal`. The changes in this PR walk the dependency tree, checking child dependencies of any Astro packages it finds recursively. It bails out of a branch when it encounters a non-Astro package. This should still be reasonably performant, while being much more robust.

## Testing

Help would be appreciated. @natemoo-re’s original PR for this code (https://github.com/withastro/astro/pull/2665) mentions:

> Not sure there's a great way to test this.

What I did manually:

1. `pnpm build` in this repo
2. `cd packages/astro && npm pack` to get a tarball for the new version
3. Create a new minimal Astro project locally
4. Install `astro-embed` and use its `YouTube` component:
    ```astro
    ---
    import { YouTube } from "astro-embed";
    ---
    <YouTube id="https://youtu.be/xtTy5nKay_Y" />
    ```
5. `npm start` — encounters errors as reported in #4071
6. `npm i ../astro/packages/astro/astro-1.1.5.tgz` to install Astro from the tarball
7. `npm start` — errors go away! ✨ 

## Docs

Bug fix for an issue that we’re not documenting in detail yet, although this is related to https://github.com/withastro/docs/issues/1455.

### CODE REVIEW

**Review by bluwy — APPROVED** [review]
This make sense to me. We have the same issue for vite-plugin-svelte (`.svelte` files and `.astro` files are exported raw), and this recursive technique is what we came up too. Reference: [dependencies.ts](https://github.com/sveltejs/vite-plugin-svelte/blob/62b9edfe0a649258809c723e0db6bf6947a1072b/packages/vite-plugin-svelte/src/utils/dependencies.ts#L10) and [usage](https://github.com/sveltejs/vite-plugin-svelte/blob/42a2e85055790421a6d9b43130ab662e35f8ad67/packages/vite-plugin-svelte/src/utils/options.ts#L401-L439).

Admittedly this is a problem in Vite that we hadn't found a nice solution yet. Ideally nodejs ESM loaders should help when that's stable.

In my second linked code, you'll also notice that we externalize transitive deps too, to prevent transitive CJS deps reporting a Vite caveat like `"module" is not defined` error in dev mode. We could probably improve that in the future too.

**Comment by delucis** [comment]
Thanks @bluwy! I had a suspicion you’d have some good insights into this issue 💜

**Comment by matthewp** [comment]
Can a test be added?

**Comment by delucis** [comment]
> Can a test be added?

Nate scared me off with his comment, but I can try!

**Comment by delucis** [comment]
Test added! I’ve also had to switch up my fix to get the tests passing, so probably a good call to try testing this.

Our existing `package.json` look-up strategy simply loaded `node_modules/${dep}/package.json` and hoped that was the correct location. That fails in monorepo environments where `node_modules` might be elsewhere making this impossible to test (and presumably also brittle for monorepo Astro users).

I’ve switched to using `createRequire` to resolve dependencies based on [the `vite-plugin-svelte` strategy](https://github.com/sveltejs/vite-plugin-svelte/blob/62b9edfe0a649258809c723e0db6bf6947a1072b/packages/vite-plugin-svelte/src/utils/dependencies.ts#L40) @bluwy linked to.

My test uses `astro-embed` which is the community package (which I maintain) that triggered the initial bug report. This is in the @astro-community org and under the @astro-community scope on npm, the latter of which I am the admin for. Not sure if there’s a better way to test this — I guess it would be hard to create a third-party package with indirect astro dependencies within the astro monorepo?

**Comment by bluwy** [comment]
The new changes looks great!

> Not sure if there’s a better way to test this — I guess it would be hard to create a third-party package with indirect astro dependencies within the astro monorepo?

I think the current test method is fine, but if we want to avoid an external dependency, we can use the `file:` protocol from pnpm so that it hardlinks the package instead of symlink, which can emulate installing a third-party package.

So you'd have something like: `"astro-lib": "file:../path/to/lib"` instead of `"astro-lib": "workspace:*"`. Vite also [does this](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md#note-on-test-dependencies) too.

**Comment by matthewp** [comment]
We do have some tests that use local packages, like this one: https://github.com/withastro/astro/blob/main/packages/astro/test/fixtures/css-assets/package.json

But I think this method is fine for now.

**Comment by delucis** [comment]
Thanks @bluwy & @matthewp! I think ultimately it would be better to have an internal package to test this, but I’ve already sunk quite a bit of time into this one, so I’ll merge and happy for us to tackle that down the road!

**Comment by beeb** [comment]
This PR makes the linter fail now:

```
astro/packages/astro/src/core/create-vite.ts
  246:13  error  'e' is already declared in the upper scope on line 232 column 12  @typescript-eslint/no-shadow
```

when running `pnpm run lint`.
